### DISCUSSION TITLE
Bidirectional shift register.

### DISCUSSION THREAD

**Discussion by Arathok** [logisim-evolution/logisim-evolution#2116] [General]
Hey guys, is there a way to add own subcircuits as components?
Id really love to use this tool to teach students about digital electronics, and it works superb for counters. We are using the Shift Register 74194 here in the practical courses, because it can shift in both directions.
Is there any plans of adding it as component? 
Or perhaps one of you already made a component they want to share?

**Comment by maehne** [comment]
Many TTL components were added by @btb and @BunchOfBits in the past. You can have a look at the implementation of similar components to the one you are intested in in the directory [src/main/java/com/cburch/logisim/std/ttl](https://github.com/logisim-evolution/logisim-evolution/tree/main/src/main/java/com/cburch/logisim/std/ttl). It is also possible to build a sub-circuit `TTL74194` from the standard Logisim components and use its CIRC file as a library.

As our resources are limited and we don't use TTL components in our courses teaching digital design, we won't implement it ourselves. However, we are always open for reviewing/merging PRs!

**Reply by BunchOfBits** [reply]
I will see what I can do. I don't have a lot of time for this either, so don't expect anything anytime soon though

**Comment by btb** [comment]
I've started implementing the 74194. There is an existing bidirectional shift register, 74299, which I'm modeling it after. I found that there is a bug however in the ShiftRegisterData class which 74299 uses. The array that holds the register values internally uses the % operator to wrap the index into the array. this works fine when shifting right, but shifting left results in a negative index and an out of bounds exception, so the 74299 has never worked correctly.

**Comment by btb** [comment]
PR #2206

**Comment by maehne** [comment]
Closing as resolved due to merge of PR #2206.
